### ISSUE TITLE
Unix file timestamps only have one-second granularity

### ISSUE THREAD

**Issue by mellinoe** [dotnet/corefx#12403]
Currently, method such as `File.GetLastWriteTime` and `FileInfo.LastWriteTime` return timestamps which are only accurate to the nearest second on Unix. On OSX, this seems to be a system-level limitation in HFS+. On most Linux filesystems, however, millisecond-level information does seem to be available, at least via `stat`. @stephentoub recommended that we could use `utimes` rather than `utime` in the native shim in order to retrieve this information.

This also seems to affect files copied by `File.Copy`, because we copy over some timestamp metadata from the source file (to match Windows).

**Comment by TicoDePano** [comment]
On .NET Core 2.1 on Linux it gets worst, as it considers the milliseconds part when comparing files, but a copy doesn't set such a resolution.

**Comment by danmosemsft** [comment]
@TicoDePano can you clarify? I think this original Issue is fixed : milliseconds are now returned on Unix and also on macOS if APFS partition. What change remains to make?

**Comment by TicoDePano** [comment]
```
Console.WriteLine($"Runtime:`` {RuntimeInformation.FrameworkDescription}");
	for(int x = 0; x < 2; x++) {
		FileInfo input = new FileInfo(Path.Combine(config.InputDir, config.File));
		FileInfo output = new FileInfo(Path.Combine(root, config.OutputDir, config.File));
		Console.WriteLine($"Input: {input.LastWriteTime.Millisecond} ms, {input.LastWriteTime.Ticks} ticks");
		Console.WriteLine($"Output: {output.LastWriteTime.Millisecond} ms, {output.LastWriteTime.Ticks} ticks");
		if(!output.Exists || input.LastWriteTime > output.LastWriteTime) {
			if(!output.Directory.Exists) {
				output.Directory.Create();
				Console.WriteLine($"Copying {group}: directory ${output.Directory.Name} created");
			}
			input.CopyTo(output.FullName, true);
			copy = true;
			Console.WriteLine($"Copying {group}: file {output.Name} copied");
		}
	}
```

**.NET Core 2.0 results:**
Runtime: .NET Core 4.6.00001.0
Input: 0 ms, 636662151720000000 ticks
Output: 0 ms, 504911120400000000 ticks
Copying images: directory $img created
Copying images: file banner1.svg copied
Input: 0 ms, 636662151720000000 ticks
Output: 0 ms, 636662151720000000 ticks

**.NET Core 2.1 results:**
Runtime: .NET Core 4.6.26606.02
Input: 205 ms, 636662151722054140 ticks
Output: 0 ms, 504911120400000000 ticks
Copying images: directory $img created
Copying images: file banner1.svg copied
Input: 205 ms, 636662151722054140 ticks
Output: 0 ms, 636662151720000000 ticks
Copying images: file banner1.svg copied

*** Output file doesn't exist initially.

**Comment by danmosemsft** [comment]
I repro this on Ubuntu 1604 -- when we copy the file the access and modify times both drop their millisec part. The change time does not, presumably because the file system sets it. 

Certainly we should fix this. @TicoDePano do you have any interest in investigating?

**Comment by danmosemsft** [comment]
Fix is here https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Native/pal_io.c#L1271-L1293 where it explicitly sets 0 for milliseconds

It needs to access milliseconds via the macros defined here https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Native/pal_io.h#L40-L70

**Comment by danmosemsft** [comment]
This might be a servicing candidate actually since presumably it breaks up-to-date checks.

**Comment by danmosemsft** [comment]
@AndyGerlicher @cdmihai this means if someone copies a file on Linux (or macOS with the latest APFS) using .NET Core 2.1, the "destination" will look older (have older modified time) than the "source". I would definitely expect that to be causing overbuilds in MSBuild.  Have you had reports?

**Comment by danmosemsft** [comment]
@Anipik could you please fix this? It would be great to get a PR up today in case there is still runway for 2.1.3

**Comment by Anipik** [comment]
I will make it my first priority.

**Comment by danmosemsft** [comment]
@Anipik I expect you would use the macro similar ot this
https://github.com/dotnet/corefx/blob/44e5bdafb8d989a220c9cf1b94f31a64a6e4f052/src/Native/Unix/System.Native/pal_io.c#L177

**Comment by Anipik** [comment]
@danmosemsft I think the bug is in managed side too. 
Because I run this test https://github.com/dotnet/corefx/compare/master...Anipik:millisec#diff-ad4fcf7234e97e8a9898813be7db41e5R30
on master 
the output was
```
  Runtime:`` .NET Core 4.6.26710.05
  Input: 782 ms, 636668666347825176 ticks
  Output: 0 ms, 504911232000000000 ticks
  Copying  directory $UnixTest_36_d95c30ea created
  0
  Copying  file UnixTest_36_2ce649ec copied
  Finished:    System.IO.FileSystem.Tests

  === TEST EXECUTION SUMMARY ===
     System.IO.FileSystem.Tests  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.201s
  782517600
  78251760
```


while after making the change the output is 

```
  Runtime:`` .NET Core 4.6.26710.05
  Input: 258 ms, 636668664312582037 ticks
  Output: 0 ms, 504911232000000000 ticks
  Copying  directory $UnixTest_36_87d8fde7 created
  258203000
  Copying  file UnixTest_36_ae463295 copied
  Finished:    System.IO.FileSystem.Tests

  === TEST EXECUTION SUMMARY ===
     System.IO.FileSystem.Tests  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.197s
  258203700
  258203700
```

This indicates our fix is working. But we have to fix the managed side too. I am debugging into managed side now

**Comment by danmosemsft** [comment]
@TicoDePano thank you for the report. This is now fixed in master and we will consider servicing 2.1

**Comment by TicoDePano** [comment]
@danmosemsft thanks and sorry for not helping with the code. I'm not working with C/C++ a long time.

**Comment by danmosemsft** [comment]
@TicoDePano no problem, you are welcome to look at another issue if you like, most of our issues labeled up-for-grabs are c# changes.

**Comment by danmosemsft** [comment]
Reopening for 2.1 port

**Comment by danmosemsft** [comment]
### Shiproom template

#### Description
Currently, when we copy the files on Unix, the millisecond attribute of the timestamp of the new file is always set to zero. This can mean the destination of the copy looks older than the source of the copy.

#### Customer Impact
MSBuild will over-build on Linux (and macOS Mojave) which can be much slower than skipping build steps.

#### Regression?
Regression from 2.0

#### Risk
Low - just change from setting zero to setting milliseconds using existing macro

**Comment by danmosemsft** [comment]
To take this in servicing we need evidence customers are actually being impacted.

**Comment by danmosemsft** [comment]
Shiproom -- I updated template with perf numbers from @Anipik .

**Comment by vivmishra** [comment]
Approved for 2.1.5

**Comment by danmosemsft** [comment]
Leaving everything in 2.1.4 until we determine whether it will have generic fixes

**Comment by joshfree** [comment]
Moving to 2.1.5 per shiproom today

**Comment by Anipik** [comment]
So does that mean we can merge the Linked PR ?

**Comment by joshfree** [comment]
@Anipik not yet.  Likely next week - please wait for @vivmishra to let us know when the branch is open for 2.1.5 commits.

**Comment by danmosemsft** [comment]
Moving label to PR per new process

**Comment by Anipik** [comment]
we have merged this change into release so closing this issue

**Comment by karelz** [comment]
Fixed in release/2.1 branch in PR #31046. Will be part of 2.1.5 release.
